lib/Frontend/CachingUtils.cpp

```diff
@@ -39,204 +44,183 @@
-    // input it actually comes from. Maybe the solution is just not to cache
-    // any commands write output to `-`.
-    file_types::ID mainOutputType = InputsAndOutputs.getPrincipalOutputType();
-    auto addInput = [&](const InputFile &Input) {
-      if (!Input.outputFilename().empty())
-        OutputToInputMap.insert(
-            {Input.outputFilename(), {Input, mainOutputType}});
-      Input.getPrimarySpecificPaths()
-          .SupplementaryOutputs.forEachSetOutputAndType(
-              [&](const std::string &Out, file_types::ID ID) {
-                OutputToInputMap.insert({Out, {Input, ID}});
-              });
-    };
-    llvm::for_each(InputsAndOutputs.getAllInputs(), addInput);
-  }
+  auto OutputRef = CAS.getReference(**Lookup);
+  if (!OutputRef)
+    return false;
 
-  file_types::ID getOutputFileType(StringRef Path) const {
-    return file_types::lookupTypeForExtension(llvm::sys::path::extension(Path));
-  }
+  auto ResultProxy = CAS.getProxy(*OutputRef);
+  if (!ResultProxy)
+    return ResultProxy.takeError();
 
-public:
-  SwiftCASOutputBackend(ObjectStore &CAS, ActionCache &Cache, ObjectRef BaseKey,
-                        const FrontendInputsAndOutputs &InputsAndOutputs)
-      : CAS(CAS), Cache(Cache), BaseKey(BaseKey),
-        InputsAndOutputs(InputsAndOutputs) {
-    initBackend(InputsAndOutputs);
-  }
+  {
+    swift::cas::CompileJobResultSchema Schema(CAS);
+    if (Schema.isRootNode(*ResultProxy)) {
+      auto Result = Schema.load(*OutputRef);
+      if (!Result)
+        return Result.takeError();
 
-private:
-  ObjectStore &CAS;
-  ActionCache &Cache;
-  ObjectRef BaseKey;
+      if (auto Err = Result->forEachOutput(
+              [&](swift::cas::CompileJobCacheResult::Output Output) -> Error {
+                return Callback(Output.Kind, Output.Object);
+              }))
+        return Err;
 
-  StringMap<std::pair<const InputFile &, file_types::ID>> OutputToInputMap;
-  const FrontendInputsAndOutputs &InputsAndOutputs;
-};
-}
-
-namespace swift {
+      return true;
+    }
+  }
+  {
+    clang::cas::CompileJobResultSchema Schema(CAS);
+    if (Schema.isRootNode(*ResultProxy)) {
+      auto Result = Schema.load(*OutputRef);
+      if (!Result)
+        return Result.takeError();
+      if (auto Err = Result->forEachOutput(
+          [&](clang::cas::CompileJobCacheResult::Output Output) -> Error {
+            file_types::ID OutputKind = file_types::ID::TY_INVALID;
+            switch (Output.Kind) {
+            case clang::cas::CompileJobCacheResult::OutputKind::MainOutput:
+              OutputKind = file_types::ID::TY_ClangModuleFile;
+              break;
+            case clang::cas::CompileJobCacheResult::OutputKind::Dependencies:
+              OutputKind = file_types::ID::TY_Dependencies;
+              break;
+            case clang::cas::CompileJobCacheResult::OutputKind::
+                SerializedDiagnostics:
+              OutputKind = file_types::ID::TY_CachedDiagnostics;
```

Comment:
Will this be compatible between clang and swift diagnostics?

Reply:
Good question, it doesn't but it should be fine here.

The reason why it needs a compatible schema is because when clang importer loads clang module, it is loading via `-fmodule-cache-key` option and clang doesn't have knowledge of swift schema, so the `MainOutput` entry is compatible, but not diagnostics entry. It is ok as long as:
* clang importer only loads module, not the diagnostics that emitted when building the module.
* clang cache key and swift cache key will not collide, so clang cc1 will not ever load swift schema.

Reply:
Maybe we should just skip these entries if they're not used?

Reply:
They are used for `emit-pcm` job from swift-frontend. It is needed for replay compilation, but not needed for module loading (which needs a compatible schema).